crimson:: add alien blue store

liu-chunmei opened the pull request:
need change seastar/src/core/reactor.cc smp::count > 1 to smp::count >=1 before build the code, otherwise alien world can't send message to crimson since no message queue setup.
also when to build release version, need make seastar use default allocator. otherwise, seastar memory allocator can't allocate enough pages for bluestore.

The main changes in this PR:
1. some code coexist in crimson seastar environment and posix environment, so add namespace to avoid same function conflict, for example  add namespace for CephContext, since the new namespace for classic ceph-osd, need modify all files declare use CephContext by including "common_fwd.h" which defined the namespace for each environment. 
2. adds an alien store wrapper for ceph blue store, which will create a thread pool, so crimson-osd can call bluestore API which is running in Posix thread from seastar thread. 
3. adjust osd stop steps, since futurized collection and bluestore destroy can't be called in osd destructor. 

. 

Signed-off-by: chunmei Liu <chunmei.liu@intel.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

liu-chunmei commented:
@tchaikov @athanatos please review alien blue store. will rebase it later.  current cyan_store and alien_store can't build together. dout.h and Context.h has conflict in seastar env and alien env, maybe we can solve it like CephContext add more namespace, but it will influence more files.

athanatos commented:
I left a few comments in an email from a few weeks ago I think should be addressed first:

I'm worried that the way the name spacing is done will be confusing in the future.  For instance, the crimson specific compilation variant (if I'm understanding this correctly) would have a ceph::common:: prefix. I think we should instead adopt the following convention for any file/function/class which compiles differently between classic and crimson.
1) No un-namespaced symbols.
2) All crimson code should be switched from ceph:: to crimson::
3) All files which can compile both ways should include a single file with a #define for a namespace symbol which can be used at the top of the file as:

`namespace TOP_LEVEL_NAMESPACE::`

This way, the behavior of all of the effected files can be controlled from a single header based on the WITH_SEASTAR/WITH_ALIEN defines by setting TOP_LEVEL_NAMESPACE to ceph:: or crimson::.
-Sam

athanatos commented:
Also, I think we need to get both cyan_store and alien_store compiling together before we have a good sense of how invasive the change will be.

tchaikov commented:
> I'm worried that the way the name spacing is done will be confusing in the future. For instance, the crimson specific compilation variant (if I'm understanding this correctly) would have a ceph::common:: prefix. I think we should instead adopt the following convention for any file/function/class which compiles differently between classic and crimson.
> 
> 1. No un-namespaced symbols.
> 2. All crimson code should be switched from ceph:: to crimson::
> 3. All files which can compile both ways should include a single file with a #define for a namespace symbol which can be used at the top of the file as:
> 

i am posting #31069 to prepare for the namespace change.

athanatos commented:
Here's an example of what I'm talking about:

https://github.com/athanatos/ceph/commit/f26269bc75cfc37130001a895c5fe32ade2a4746

That patch introduces a ceph_crimson_common header which defines a TOPNSPC macro (among others) which ceph_crypto then uses instead of specifically \#ifdef'ing each and every namespace or method name.

Note that users may use TOPNSPC if they also can compile in both contexts or the namespace name directly otherwise.  The idea would be to adapt the other files (like ceph_context) in the same way.  The alien:: case should also hopefully solve the cyanstore/alienstore problem as well.

athanatos commented:
If it's not clear how to proceed, let me know -- I'll adapt a few more files to make it clearer.

liu-chunmei commented on src/CMakeLists.txt:
```diff
@@ -512,6 +490,28 @@ add_subdirectory(os)
 
 add_subdirectory(osd)
 
+if(WITH_SEASTAR)
```
crimson-os need link other object, need them build before crimson.

liu-chunmei commented:
@athanatos Hi Sam, it is a good idea for the class which has the same implementation only need a different nampspace name.  but for different implement, how can we make it easily? such as 
#ifdef WITH_SEASTAR
  crimson::common::ConfigProxy& conf(CephContext*) {
    return ceph::common::local_conf();
  }
#else
  ConfigProxy& conf(CephContext* cct) {
    return cct->_conf;
  }
#endif
Thanks!

athanatos commented:
Yes, you'll need two definitions.  I suggest you put them in the same file with the namespace names I've defined in ceph_crimson_common.h.  If you do it that way, other users need only include (in this case) the single config_proxy.h and use TOPNSPC:: to resolve the correct implementation.  I can work up an example tomorrow if that's not clear -- let me know.

liu-chunmei commented:
@athanatos @tchaikov @rzarzynski PR is updated.

tchaikov commented on src/common/ceph_context.cc:
```diff
@@ -53,7 +53,8 @@ using ceph::HeartbeatMap;
 #include <pthread.h>
 
-#ifdef WITH_SEASTAR
+#if defined (WITH_SEASTAR) && !defined (WITH_ALIEN)
```
nit, please remove the space after `defined`.

tchaikov commented on src/common/ceph_context.h:
```diff
@@ -42,12 +43,15 @@
 class CryptoRandom;
 
+namespace TOPNSPC::common {
+  class CephContextServiceThread;
```
this class is defined in `ceph::common`, see `ceph_context.cc`. or am i missing something?

tchaikov commented on src/common/ceph_context.cc:
```diff
@@ -356,6 +360,7 @@ class LogObs : public md_config_obs_t {
 };
 
 
+namespace TOPNSPC::common {
```
why?

liu-chunmei replied:
chage it to ceph::common,  need add the namespace so friend class can access private member. such as _cct->_heartbeat_map works.

athanatos commented on src/include/ceph_crimson_common.h:
```diff
@@ -0,0 +1,24 @@
+
+#if defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
+#define TOPNSPC crimson
+#define CRIMSONENV
```
Yeah, I meant that as an example of how we'd allow source files to do #ifdef's based on environment, no need to include them prior to actual users though.

liu-chunmei replied:
remove it.

athanatos commented on src/include/cephfs/libcephfs.h:
```diff
@@ -28,6 +28,7 @@
 #include <fcntl.h>
 
 #include "ceph_statx.h"
+#include "include/ceph_crimson_common.h"
```
Yeah, it shouldn't end up in public header.

liu-chunmei replied:
in libcephfs.h declare use CephContext, we have already put CephContext under TOPNSPC::common::CephContext.  so every palce uses CephContext need change.

liu-chunmei replied:
I will remove include "ceph_crimson_common.h" and declare ceph::common:CephContext under cplusplus instead.

liu-chunmei commented on src/common/ceph_context.h:
```diff
@@ -83,7 +88,11 @@ class CephContext {
 };
+}
 #else
+#ifdef __cplusplus
```
there is a unit test test_c_headers, will fail if there is namespace not under cplusplus macro.

liu-chunmei commented on src/include/ceph_crimson_common.h:
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#ifdef __cplusplus
```
yes,  test_c_headers.c include libcephfs.h, libcephfs.h include ceph_crimson_common.h.  but I can move cplusplus check in libcephfs.h.  will remvo it here.

liu-chunmei commented on src/kv/LevelDBStore.h:
```diff
@@ -31,8 +31,6 @@
 // reinclude our assert to clobber the system one
 # include "include/ceph_assert.h"
 
-class PerfCounters;
```
sure, will change all places where mssing ceph_crimson_common.h

liu-chunmei commented on src/include/ceph_crimson_common.h:
```diff
@@ -0,0 +1,24 @@
+#pragma once
```
@athanatos do you agree the name?

tchaikov commented:
@liu-chunmei could you rebase the change and address the comments?

liu-chunmei commented:
@tchaikov PR is updated, but not rebase, after most review done, I will rebase it. otherwise maybe the other PR merged will influence it and need rebase again, since so many files are touched.  fio test works now.

cyx1231st commented on src/crimson/os/alienstore/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+  target_link_libraries(crimson-alien ${AIO_LIBRARIES})
+endif(HAVE_LIBAIO)
+
+  target_compile_definitions(crimson-alien PRIVATE -DWITH_SEASTAR -DWITH_ALIEN)
```
Not sure, but I see each `WITH_ALIEN` is associated with `WITH_SEASTAR` with definition `#if defined (WITH_SEASTAR) && !defined (WITH_ALIEN)`. It seems to me that define both `WITH_SEASTAR` and `WITH_ALIEN` is just identical to not define `WITH_SEASTAR`...

So why do we need `WITH_ALIEN`?

liu-chunmei replied:
alien blue store built under crimson, all crimson code built under WITH_SEASTAR=1, so alien blue store with default definition WITH_SEASTAR=1, but these code should run in posix thread, should make it not go to WITH_SEASTAR branch, so add another option to identify it.

liu-chunmei commented:
@tchaikov @athanatos @rzarzynski @cyx1231st  PR is rebased, please review ASAP.

liu-chunmei commented:
@tchaikov update the commit message, please review, if you still have questions, you can leave comments or call me to discuss.

liu-chunmei commented on src/perfglue/CMakeLists.txt:
```diff
@@ -1,8 +1,8 @@
-if(ALLOCATOR STREQUAL "tcmalloc")
+if(ALLOCATOR STREQUAL "tcmalloc" AND NOT WITH_SEASTAR)
```
I got the following error before crimson-osd start, (seastar said default allocator can't use heap profiler). 
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
src/tcmalloc.cc:283] Attempt to free invalid pointer 0x604000000010
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff3282801 in __GI_abort () at abort.c:79
#2  0x00007ffff4fb9fb9 in tcmalloc::Log(tcmalloc::LogMode, char const*, int, tcmalloc::LogItem, tcmalloc::LogItem, tcmalloc::LogItem, tcmalloc::LogItem) ()
   from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#3  0x00007ffff4fadd49 in ?? () from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#4  0x00007ffff4fcd6b9 in tc_free ()
   from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#5  0x00007ffff4fc2720 in MallocExtension::Initialize() ()
   from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#6  0x00007ffff4face0a in ?? () from /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
#7  0x00007ffff7de5733 in call_init (env=0x7fffffffe1b8, argv=0x7fffffffe138,
    argc=15, l=<optimized out>) at dl-init.c:72
#8  _dl_init (main_map=0x7ffff7ffe170, argc=15, argv=0x7fffffffe138,
    env=0x7fffffffe1b8) at dl-init.c:119
#9  0x00007ffff7dd60ca in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#10 0x000000000000000f in ?? ()
#11 0x00007fffffffe44b in ?? ()
#12 0x00007fffffffe474 in ?? ()
#13 0x00007fffffffe477 in ?? ()
#14 0x00007fffffffe479 in ?? ()
#15 0x00007fffffffe47c in ?? ()
#16 0x00007fffffffe49f in ?? ()
#17 0x00007fffffffe4a6 in ?? ()
#18 0x00007fffffffe4ac in ?? ()
#19 0x00007fffffffe4d5 in ?? ()
#20 0x00007fffffffe4e0 in ?? ()
#21 0x00007fffffffe505 in ?? ()
#22 0x00007fffffffe50b in ?? ()
#23 0x00007fffffffe50d in ?? ()
#24 0x00007fffffffe516 in ?? ()
#25 0x00007fffffffe518 in ?? ()
#26 0x0000000000000000 in ?? ()

tchaikov replied:
could you move this change into a separated commit. and put the rationales in the commit message?

liu-chunmei commented on src/crimson/os/alienstore/alien_collection.h:
```diff
@@ -0,0 +1,32 @@
+    auto s = static_cast<AlienStore*>(store);
+    return  s->tp.submit([this](){
+      sch.reset();
+      return 0;
```
threadpool.submit  can not accept void return,  see the following error: 
usr/include/c++/9/tuple:165:13: error: invalid use of ‘void’
/usr/include/c++/9/tuple: In instantiation of ‘struct std::_Tuple_impl<0, void>’:
/usr/include/c++/9/tuple:523:11:   required from ‘class std::tuple<void>’
/usr/include/c++/9/type_traits:1235:12:   required from ‘struct std::is_trivially_destructible<std::tuple<void> >’
/home/chunmei/ceph/src/seastar/include/seastar/core/future.hh:189:51:   required from ‘constexpr const bool seastar::internal::can_inherit<std::tuple<void> >’
/home/chunmei/ceph/src/seastar/include/seastar/core/future.hh:196:44:   required from ‘struct seastar::internal::uninitialized_wrapper<std::tuple<void> >’
/home/chunmei/ceph/src/seastar/include/seastar/core/future.hh:346:8:   required from ‘struct seastar::future_state<void>’
/home/chunmei/ceph/src/crimson/thread/ThreadPool.h:28:28:   required from ‘struct crimson::thread::Task<const crimson::thread::ThreadPool::submit(Func&&, Args&& ...) [with Func = crimson::os::AlienCollection::clear(crimson::os::FuturizedStore*)::<lambda()>; Args = {}]::<lambda()>, void>’

tchaikov replied:
ack. will need to revise the ThreadPool to address this.. could you add a comment to briefly explain the reason here?

liu-chunmei replied:
template<typename Func, typename T = std::invoke_result_t<Func>>
struct Task final : WorkItem {
  Func func;
  seastar::future_state<T> state;
in ThreadPool.h  need function result type T to define seastar::future_state<T> state, if it is void, has this build error.

liu-chunmei commented on src/crimson/os/alienstore/alien_store.cc:
```diff
@@ -0,0 +1,367 @@
+  store = new BlueStore(cct.get(), path);
+
+}
+/*
```
I didn't find the proper place to call alienstore->start() in crismon-osd startup flow. any suggestion here?

liu-chunmei commented on src/crimson/os/alienstore/alien_store.cc:
```diff
@@ -0,0 +1,367 @@
+  return tp.submit([=](){
+    std::string value(4096, '\0');
+    int r = store->read_meta(key, &value);
+    if (r >0) {
```
r is used later for return pair, so keep it here.

liu-chunmei commented:
@tchaikov updated according to your comments, please review it again. Thanks!

tchaikov commented on src/crimson/osd/osd.cc:
```diff
@@ -341,24 +341,28 @@ seastar::future<> OSD::_add_me_to_crush()
     return seastar::now();
   }
-  const double weight = [this] {
+  auto get_weight = [this] () {
```
nit, /()// as if a lambda does not accepts any parameters the parameter list can be dispensed.

liu-chunmei commented on src/crimson/osd/pg_backend.h:
```diff
@@ -45,7 +45,6 @@ class PGBackend
 					   const ec_profile_t& ec_profile);
 
   using read_errorator = ll_read_errorator::extend<
-    crimson::ct_error::input_output_error,
```
read_errorator is changed in futurized store, crimson::ct_error::input_output_error is add in futurized store since bluestore may return this error. so remove it here.

tchaikov commented on src/crimson/osd/osd.cc:
```diff
@@ -341,24 +341,28 @@ seastar::future<> OSD::_add_me_to_crush()
-      return std::max(.00001, double(total) / double(1ull << 40)); // TB
+       return store->stat().then([](auto st) {
+         auto total = st.total;
+	 return seastar::make_ready_future<double>(std::max(.00001, double(total) / double(1ull << 40))); // TB
```
wrap at 80. and might want to be consistent when using space and tab for indent.

tchaikov commented on src/crimson/osd/osd.cc:
```diff
@@ -341,24 +341,28 @@ seastar::future<> OSD::_add_me_to_crush()
-  }})", whoami, weight, loc);
-  return monc->run_command({cmd}, {}).then(
+  };
+  return get_weight().then([this](auto weight) {
```
why not just chain the calls? i.e. remove the `get_weight` local variable.

liu-chunmei replied:
this logic looks more clear than chain them together.

tchaikov commented on src/crimson/osd/osd_meta.cc:
```diff
@@ -25,7 +25,7 @@ seastar::future<bufferlist> OSDMeta::load_map(epoch_t e)
                      osdmap_oid(e), 0, 0,
                      CEPH_OSD_OP_FLAG_FADVISE_WILLNEED).handle_error(
-    crimson::ct_error::enoent::handle([e] {
+    read_errorator::all_same_way([e] {
```
@liu-chunmei why is this change a part of this commit? and why do we need this change?

tchaikov commented on src/common/ceph_context.h:
```diff
@@ -30,7 +30,7 @@
 #include "common/cmdparse.h"
 #include "common/code_environment.h"
-#ifdef WITH_SEASTAR
+#if  defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
```
remove the redundant space.

tchaikov commented on src/common/config_proxy.h:
```diff
@@ -124,6 +124,11 @@ class ConfigProxy {
     return &values;
   }
+#ifdef WITH_SEASTAR
+  void set_config_values(ConfigValues* val) {
```
can we pass a `const ConfigValues&` instead? if `val` is always valid pointer.

tchaikov commented on src/crimson/os/futurized_store.cc:
```diff
@@ -1,14 +1,17 @@
 std::unique_ptr<FuturizedStore>
 FuturizedStore::create(const std::string& type,
-                       const std::string& data)
+                       const std::string& data, ConfigValues* values)
```
why not pass by const reference? like `const ConfigValues& values` ?

tchaikov commented on src/crimson/os/futurized_store.h:
```diff
@@ -27,7 +27,8 @@ class FuturizedStore {
   static std::unique_ptr<FuturizedStore> create(const std::string& type,
-                                                const std::string& data);
+                                                const std::string& data,
+						ConfigValues* values);
```
ditto.

tchaikov commented on src/global/global_context.cc:
```diff
@@ -23,7 +23,7 @@
 CephContext *g_ceph_context = NULL;
 ConfigProxy& g_conf() {
-#ifdef WITH_SEASTAR
+#if defined (WITH_SEASTAR) && !defined (WITH_ALIEN)
```
please remove the space after `defined`.

tchaikov commented on src/include/Context.h:
```diff
@@ -31,6 +31,7 @@
 /*
  * GenContext - abstract callback class
  */
+
```
please drop this empty line.

tchaikov commented on src/crimson/os/cyanstore/cyan_store.cc:
```diff
@@ -416,6 +416,7 @@ seastar::future<> CyanStore::do_transaction(CollectionRef ch,
       t.get_on_applied_sync()}) {
     if (i) {
       i->complete(0);
+      delete i;
```
why?

liu-chunmei replied:
alien store need keep context alive (for alienstore OnCommit promise get_future),  so in Context.h  comment out delete this,  if I identify cyan_store from alien store, I need add more namespace, so here treat them the same. 
#ifndef WITH_SEASTAR
    delete this;
#endif

athanatos replied:
Wait, so with alien, Context::complete doesn't delete the context?  I think that's incredibly dangerous.  If anyone modifies BlueStore to use a context for a purpose other than the exposed interface, it'll introduce a memory leak.  Is there really no better way to do this?  Why can't you wrap the context before passing it to bluestore?

liu-chunmei commented:
@tchaikov  blue store can return two errors, enoent, and input_output_error, (which is defined in read_errorator in futirized_store.h), so change the following handle. 
    crimson::ct_error::enoent::handle([e] {
    read_errorator::all_same_way([e] {

liu-chunmei commented:
@tchaikov updated.

tchaikov commented:
@liu-chunmei

- i'd suggest put more details in the commit message. you could move the https://github.com/ceph/ceph/pull/31041#issue-330754790 into the commit message of related commit
- i add a commit to reformat alien_store.cc -- there are too many of them. please fold it into the original commit, if you are fine with it.
- the build fails in my testbed. like
```
In file included from ../src/global/global_context.cc:20:
../src/crimson/common/config_proxy.h: In lambda function:
../src/crimson/common/config_proxy.h:72:52: error: no match for ‘operator[]’ (operand types are ‘ObserverMgr<ceph::md_config_obs_impl<crimson::common::ConfigProxy> >::rev_obs_map’ {aka ‘std::map<ceph::md_config_obs_impl<crimson::common::ConfigProxy>*, std::set<std::__cxx\
11::basic_string<char> >, std::less<ceph::md_config_obs_impl<crimson::common::ConfigProxy>*>, std::allocator<std::pair<ceph::md_config_obs_impl<crimson::common::ConfigProxy>* const, std::set<std::__cxx11::basic_string<char> > > > >’} and ‘md_config_obs_t*’ {aka ‘ceph::md\
_config_obs_impl<ceph::common::ConfigProxy>*’})
   72 |                                             rev_obs[obs].insert(key);
      |                                                    ^
In file included from /usr/include/c++/9/map:61,
                 from ../src/common/ceph_context.h:19,
                 from ../src/global/global_context.cc:18:
/usr/include/c++/9/bits/stl_map.h:490:7: note: candidate: ‘std::map<_Key, _Tp, _Compare, _Alloc>::mapped_type& std::map<_Key, _Tp, _Compare, _Alloc>::operator[](const key_type&) [with _Key = ceph::md_config_obs_impl<crimson::common::ConfigProxy>*; _Tp = std::set<std::__c\
xx11::basic_string<char> >; _Compare = std::less<ceph::md_config_obs_impl<crimson::common::ConfigProxy>*>; _Alloc = std::allocator<std::pair<ceph::md_config_obs_impl<crimson::common::ConfigProxy>* const, std::set<std::__cxx11::basic_string<char> > > >; std::map<_Key, _Tp\
, _Compare, _Alloc>::mapped_type = std::set<std::__cxx11::basic_string<char> >; std::map<_Key, _Tp, _Compare, _Alloc>::key_type = ceph::md_config_obs_impl<crimson::common::ConfigProxy>*]’
  490 |       operator[](const key_type& __k)
      |       ^~~~~~~~
/usr/include/c++/9/bits/stl_map.h:490:34: note:   no known conversion for argument 1 from ‘md_config_obs_t*’ {aka ‘ceph::md_config_obs_impl<ceph::common::ConfigProxy>*’} to ‘ceph::md_config_obs_impl<crimson::common::ConfigProxy>* const&’
  490 |       operator[](const key_type& __k)
      |                  ~~~~~~~~~~~~~~~~^~~
```
- please address the build warnings like
```
../src/crimson/os/alienstore/alien_store.cc: In constructor ‘crimson::os::AlienStore::AlienStore(const string&, const ConfigValues&)’:
../src/crimson/os/alienstore/alien_store.cc:58:13: warning: ignoring returned value of type ‘seastar::future<>’, declared with attribute nodiscard [-Wunused-result]
   58 |   tp.start();
      |             ^
```

and please avoid using the C style cast in alien_store.cc, we could revise `ObjectStore::omap_get_values()` in another commit.

tchaikov commented on src/crimson/os/alien_store.cc:
```diff
@@ -0,0 +1,345 @@
+  logger().debug("{}", __func__);
+  return seastar::do_with(ceph::bufferlist{}, [=](auto &bl){
+    return tp.submit([=, &bl](){
+      auto c =static_cast<AlienCollection*>(ch.get());
```
@liu-chunmei please address this comment. i fixed most of them.

tchaikov commented on src/include/Context.h:
```diff
@@ -75,7 +76,9 @@ class Context {
     finish(r);
+#if !defined(WITH_SEASTAR) || (defined(WITH_SEASTAR) && !defined(WITH_BLUESTORE))
     delete this;
+#endif
```
@liu-chunmei could you add a comment to address this comment. i also had trouble understanding this.

liu-chunmei replied:
this is outdated,  currently it is #ifndef WITH_SEASTAR. in AlienStore, I need keep callback function OnCommit alive because there is a promise in it to set value so alien store can get_future to invoke to do continuation.  if delete itself in callback function ( at here), alien store continuation can't woken up.  if I identify cyanstore and alienstore, need add namespace for these callback context, so I treat cyanstore and alienstore the same. delete it in store functions.

liu-chunmei commented:
@tchaikov  updated, please review.

tchaikov commented:
```
10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 11.4302/18.8147  => rejected
10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 2926.0/4816.0  => rejected
10:40:41 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 76.3741/219.132  => accepted
10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00545813/0.00331131  => rejected
10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 11.5107/18.4793  => rejected
10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 2946.0/4730.0  => rejected
10:40:41 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 60.8523/190.253  => accepted
10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00542472/0.00335986  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 47.9852/61.1405  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 12284.0/15651.0  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 1170.24/507.93  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.0012937/0.00101602  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 48.3459/60.9611  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 12376.0/15606.0  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 1215.88/323.636  => rejected
10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00128358/0.00101952  => rejected
10:40:41 - WARNING  - cbt      - 14 tests failed out of 16
Build step 'Execute shell' marked build as failure
```

liu-chunmei commented:
> and please avoid using the C style cast in alien_store.cc, we could revise 
I think I fix the C stly cast in alien_store.cc, do I miss some one? can you point it out?

>ObjectStore::omap_get_values() in another commit.
misunderstand your meaning here, I move it from one commit to another commit, not a new commit, since you have do it, I will keep it.
> 
> @liu-chunmei could you please reconsider this comment?


> also, i
> 
> added a bunch of cleanups on top of your change. could you review these change and squash them if appropriate? as in #32742, we now allow the alien to  future<>.
will do it. 

> moved the change to bluestore into a separated commit.
> used using in admin_socket.cc.
> added dout_context defines to silence GCC warnings
> compiled xxhash.c to fix the link failure.

liu-chunmei commented:
> 10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 11.4302/18.8147  => rejected
> 10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 2926.0/4816.0  => rejected
> 10:40:41 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 76.3741/219.132  => accepted
> 10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00545813/0.00331131  => rejected
> 10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 11.5107/18.4793  => rejected
> 10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 2946.0/4730.0  => rejected
> 10:40:41 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 60.8523/190.253  => accepted
> 10:40:41 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00542472/0.00335986  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 47.9852/61.1405  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 12284.0/15651.0  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 1170.24/507.93  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.0012937/0.00101602  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 48.3459/60.9611  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 12376.0/15606.0  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 1215.88/323.636  => rejected
> 10:40:41 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00128358/0.00101952  => rejected
> 10:40:41 - WARNING  - cbt      - 14 tests failed out of 16
> Build step 'Execute shell' marked build as failure

@tchaikov  what is this test, how do I do this test?

liu-chunmei commented:
@tchaikov updated, please review.

tchaikov commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -1757,6 +1757,8 @@ void BlueStore::OnodeSpace::dump(CephContext *cct)
 
 #undef dout_prefix
 #define dout_prefix *_dout << "bluestore.sharedblob(" << this << ") "
+#undef dout_context
```
i'd suggest keep my Signed-off-by in this commit.

tchaikov commented on src/os/ObjectStore.h:
```diff
@@ -705,6 +705,15 @@ class ObjectStore {
     std::map<std::string, ceph::buffer::list> *out ///< [out] Returned keys and values
     ) = 0;
 
+#ifdef WITH_SEASTAR
```
could you move this change into "os/bluestore: add BlueStore::omap_get_values()"?

tchaikov commented on src/crimson/osd/osd_meta.h:
```diff
@@ -31,6 +32,9 @@ class OSDMeta {
     : store{store}, coll{coll}
   {}
 
+  seastar::future<> clear() {
```
nit, i'd suggest move this change into a separated change. and explain why we want to clear the OSDMeta in the commit message.

tchaikov commented on src/crimson/os/CMakeLists.txt:
```diff
@@ -2,6 +2,12 @@ add_library(crimson-os
+
+if(WITH_BLUESTORE)
+  add_subdirectory(alienstore)
+endif(WITH_BLUESTORE)
```
please remove "WITH_BLUESTORE" in the endif() line.

tchaikov commented on src/crimson/os/alienstore/CMakeLists.txt:
```diff
@@ -0,0 +1,65 @@
+     ${PROJECT_SOURCE_DIR}/src/os/bluestore/aio.cc)
+endif()
+
+add_library(crimson-alien STATIC ${crimson_alien_srcs}
```
s/crimson-alien/crimson-alienstore/

tchaikov commented:
> @tchaikov what is this test, how do I do this test?

@liu-chunmei https://github.com/ceph/ceph/blob/master/doc/dev/crimson.rst#cbt-based-testing

tchaikov commented:
@liu-chunmei just a couple nits, i think we are close.

tchaikov commented on src/crimson/os/alienstore/alien_store.cc:
```diff
@@ -0,0 +1,363 @@
+    return tp.submit([=, &aset] {
+      auto c = static_cast<AlienCollection*>(ch.get());
+      return store->getattrs(c->collection, oid,
+		             reinterpret_cast<map<string,bufferptr>&>(aset));
```
```c++
using attrs_t = std::map<std::string, ceph::bufferptr, std::less<>>;
```

actually i suggest add `ObjectStore` to use map with transparent comparator. so we can avoid this `reinterpret_cast `. but we can leave this for another PR.

tchaikov replied:
i am working on this. seems it's non trivial.

tchaikov replied:
@liu-chunmei JFYI, my endeavours posted at #32783

liu-chunmei commented:
@tchaikov update, please review.  I am still trying to do cbt test now. when you build the code, need pay attention : 
1. need change seastar/src/core/reactor.cc smp::count > 1 to smp::count >=1 before build the code, otherwise alien world can't send message to crimson since no message queue setup.
2. also when to build release version, need make seastar use default allocator. otherwise, seastar memory allocator can't allocate enough pages for bluestore.

tchaikov commented:
> need change seastar/src/core/reactor.cc smp::count > 1 to smp::count >=1 before build the code, otherwise alien world can't send message to crimson since no message queue setup.

this is expected. as alien machinery uses the smp queue to pass the message back and forth. guess we will have to use multi-reactor to run crimson-osd.

tchaikov commented:
@liu-chunmei i added some cleanup commits. please review and squash if they look good to you.

liu-chunmei commented:
@tchaikov updated, please review.

tchaikov commented on src/crimson/osd/CMakeLists.txt:
```diff
@@ -31,6 +31,7 @@ add_executable(crimson-osd
   ${PROJECT_SOURCE_DIR}/src/osd/PGLog.cc
   ${PROJECT_SOURCE_DIR}/src/osd/osd_perf_counters.cc
   )
+target_compile_definitions(crimson-osd PRIVATE -DWITH_SEASTAR -DWITH_BLUESTORE)
```
if you always enable `WITH_BLUESTORE`, why bother checking it in `src/crimson/os/CMakeLists.txt` ?

also, `-DWITH_SEASTAR` will be added when linked against libcrimson as it's part of the public interface of `crimson` exposed via `crimson::cflags`. so there is no need to specify it. am i missing anything?

tchaikov commented on src/include/Context.h:
```diff
@@ -75,7 +75,9 @@ class Context {
   virtual ~Context() {}       // we want a virtual destructor!!!
   virtual void complete(int r) {
     finish(r);
+#ifndef WITH_SEASTAR
```
could you add a comment in the code to explain why?

tchaikov commented:
@liu-chunmei ping?

liu-chunmei commented:
@tchaikov @athanatos @rzarzynski  please review. 
kefu, need modify seastar to let alien store works. like the following patch.
1. use default allocator for release version.
2. when smp =1, let that check satisfied to set up queue for alien word.

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 924a6ea3..f6ace2d1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -898,6 +898,12 @@ if ((CMAKE_BUILD_TYPE STREQUAL "Dev") OR (CMAKE_BUILD_TYPE STREQUAL "Debug"))
       SEASTAR_TYPE_ERASE_MORE)
 endif ()

+if (CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
+  target_compile_definitions (seastar
+    PUBLIC
+      SEASTAR_DEFAULT_ALLOCATOR)
+endif ()
+
 target_compile_definitions (seastar
   PRIVATE ${Seastar_PRIVATE_COMPILE_DEFINITIONS})

diff --git a/src/core/reactor.cc b/src/core/reactor.cc
index b5a7f543..fa7c11ea 100644
--- a/src/core/reactor.cc
+++ b/src/core/reactor.cc
@@ -2554,7 +2554,7 @@ int reactor::run() {
     // 5. kernel submission: for I/O, will submit what was generated from last step.
     // 6. reap kernel events completion: some of the submissions from last step may return immediately.
     //                                   For example if we are dealing with poll() on a fd that has events.
-    if (smp::count > 1) {
+    if (smp::count >= 1) {
         smp_poller = poller(std::make_unique<smp_pollfn>(*this));
     }

athanatos commented:
I'm confused by that seastar change.  Under what conditions would smp::count be < 1?

athanatos commented on src/crimson/os/alienstore/alien_store.cc:
```diff
@@ -0,0 +1,367 @@
+  void finish(int) override {
+    auto fut = seastar::alien::submit_to(cpuid,[this]{
+      if (on_commit){
+        on_commit->complete(0);
```
AFAIK, bluestore cannot report an error on the commit path.

liu-chunmei commented:
> I'm confused by that seastar change.  Under what conditions would smp::count be < 1?

hi Sam,  it never <1. it should be >=1 any time. for our case it =1. but the original code only check smp::count >1 then setup the communication queue, for our case,  we need make it setup queue for smp::count =1.

athanatos commented:
Right, so you should probably remove the branch entirely.

liu-chunmei commented:
> Right, so you should probably remove the branch entirely.

sure.

athanatos commented on src/crimson/os/alienstore/alien_store.cc:
```diff
@@ -0,0 +1,380 @@
+        return tp->submit([=, &txn, &callback] {
+          txn.register_on_commit(callback.get());
+          auto c = static_cast<AlienCollection*>(ch.get());
+          return store->queue_transaction(c->collection, std::move(txn));
```
Here, you're directly queuing the Transaction with it's nested Contexts.  Why not pull out those contexts and wrap them in something that bluestore can safely use and delete?

liu-chunmei replied:
I need a complete context to inform crimson bluestore finished the process then the continuation can go on. so the context should keep alive and can't be deleted in bluestore. That is the simply way I can get.  could you please give me some suggestion what mechanism we can use here?

athanatos replied:
I missed where you are already wrapping them in the OnCommit callback.

I think something like https://github.com/athanatos/ceph/tree/sjust/ceph_seastar_alien_blue_store might solve the problem -- it allocates the context in bluestore's thread so that bluestore will only release the context itself -- everything else gets released from the crimson reactor.

What vstart.sh command and config file are you using to test?

liu-chunmei replied:
the following is the cmmmand I used.  config fie is vstart default 
MDS=0 MGR=1 OSD=1 MON=1 ../src/vstart.sh -n --without-dashboard --bluestore -X  --crimson
bin/ceph osd pool create rbd 128
bin/rbd create fio_test --size 10G --image-format=2 --rbd_default_features=3
LD_LIBRARY_PATH=lib/ fio ../rbd_write.fio

liu-chunmei replied:
@athanatos I have committed your patch in this PR.

athanatos commented:
@liu-chunmei Does it work for you?  I can't get crimson to start on this branch with or without those patches at the moment.  What arguments are you giving to do_cmake?

liu-chunmei commented:
> @liu-chunmei Does it work for you?  I can't get crimson to start on this branch with or without those patches at the moment.  What arguments are you giving to do_cmake?

yes, I just finish build and test with OSD=3,  ./do_cmake.sh -DWITH_SEASTAR=ON
BTW, you have to modify seastar/src/core/reactor.cc. (smp::count >1) if you fond it can't finish mkfs command.  and libseastar.a failed when build, you have to add 
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 924a6ea3..66c41427 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -298,6 +298,7 @@ seastar_find_dependencies ()

 # Private build dependencies not visible to consumers
 find_package (ragel 6.10 REQUIRED)
+find_package (Threads REQUIRED)

 #
 # Code generation helpers.
@@ -683,7 +684,8 @@ target_link_libraries (seastar
     lksctp-tools::lksctp-tools
     protobuf::libprotobuf
     rt::rt
-    yaml-cpp::yaml-cpp)
+    yaml-cpp::yaml-cpp
+    Threads::Threads)

athanatos commented:
@liu-chunmei Can we add a patch with those changes to our submodule for the moment?

tchaikov commented:
the seastar change has been merged upstream and the submodule was also updated to pick up that change . see c3cbc2501f112e83e848258456a4e53e45198629

@liu-chunmei you might want to rebase your change to master.

liu-chunmei commented:
> the seastar change has been merged upstream and the submodule was also updated to pick up that change . see c3cbc25
> @liu-chunmei you might want to rebase your change to master.

@tchaikov rebased, updated, please review.

tchaikov commented:
jenkins test make check

tchaikov commented:
- http://pulpito.ceph.com/kchai-2020-02-29_17:36:34-rados-wip-kefu-testing-2020-02-29-2106-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-02_05:23:09-rados-wip-kefu-testing-2020-03-01-2348-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-02_05:23:09-rados-wip-kefu-testing-2020-03-01-2348-distro-basic-smithi/

tchaikov commented:
@liu-chunmei could you upstream the fix of
```diff
diff --git a/src/core/reactor.cc b/src/core/reactor.cc
index b5a7f543..3aadb9b1 100644
--- a/src/core/reactor.cc
+++ b/src/core/reactor.cc
@@ -2554,9 +2554,7 @@ int reactor::run() {
     // 5. kernel submission: for I/O, will submit what was generated from last step.
     // 6. reap kernel events completion: some of the submissions from last step may return immediately.
     //                                   For example if we are dealing with poll() on a fd that has events.
-    if (smp::count > 1) {
-        smp_poller = poller(std::make_unique<smp_pollfn>(*this));
-    }
+    smp_poller = poller(std::make_unique<smp_pollfn>(*this));

     poller reap_kernel_completions_poller(std::make_unique<reap_kernel_completions_pollfn>(*this));
     poller io_queue_submission_poller(std::make_unique<io_queue_submission_pollfn>(*this));
```

wjwithagen commented:
@liu-chunmei @tchaikov 
I suspect that this PR is the reason for an error in my Jenkins testrun.
I guess clang does not like some of the changes.
Can you suggest a fix to try?

It is full of
```
home/jenkins/workspace/ceph-master/src/mgr/MgrCap.h:12:1: error: declaration conflicts with target of using declaration already in scope
class CephContext;
^
/home/jenkins/workspace/ceph-master/src/include/common_fwd.h:10:9: note: target of using declaration
  class CephContext;
        ^
/home/jenkins/workspace/ceph-master/src/include/common_fwd.h:22:24: note: using declaration
using TOPNSPC::common::CephContext;
                       ^
In file included from /home/jenkins/workspace/ceph-master/src/mgr/MgrCap.cc:24:
/home/jenkins/workspace/ceph-master/src/mgr/MgrCap.h:136:7: error: reference to 'CephContext' is ambiguous
      CephContext *cct,
      ^
/home/jenkins/workspace/ceph-master/src/mgr/MgrCap.h:12:7: note: candidate found by name lookup is 'CephContext'
class CephContext;
      ^
/home/jenkins/workspace/ceph-master/src/include/common_fwd.h:22:24: note: candidate found by name lookup is 'CephContext'
using TOPNSPC::common::CephContext;
                       ^
In file included from /home/jenkins/workspace/ceph-master/src/mgr/MgrCap.cc:24:
/home/jenkins/workspace/ceph-master/src/mgr/MgrCap.h:184:19: error: reference to 'CephContext' is ambiguous
  bool is_capable(CephContext *cct,
                  ^
/home/jenkins/workspace/ceph-master/src/mgr/MgrCap.h:12:7: note: candidate found by name lookup is 'CephContext'
class CephContext;
      ^
/home/jenkins/workspace/ceph-master/src/include/common_fwd.h:22:24: note: candidate found by name lookup is 'CephContext'
using TOPNSPC::common::CephContext;
                       ^
In file included from /home/jenkins/workspace/ceph-master/src/mgr/MgrCap.cc:27:
In file included from /home/jenkins/workspace/ceph-master/src/common/debug.h:18:
In file included from /home/jenkins/workspace/ceph-master/src/common/dout.h:28:
/home/jenkins/workspace/ceph-master/src/global/global_context.h:24:8: error: reference to 'CephContext' is ambiguous
extern CephContext *g_ceph_context;
       ^
/home/jenkins/workspace/ceph-master/src/mgr/MgrCap.h:12:7: note: candidate found by name lookup is 'CephContext'
class CephContext;
      ^
/home/jenkins/workspace/ceph-master/src/include/common_fwd.h:22:24: note: candidate found by name lookup is 'CephContext'
using TOPNSPC::common::CephContext;
                       ^
```

tchaikov commented:
@wjwithagen ack. lemme take a look.

wjwithagen commented:
@tchaikov 
I can revert the PR locally. For a while that is, until others are starting to depend on it.
But it might give you more time to think about it.

tchaikov commented:
@wjwithagen #33657

liu-chunmei commented:
> @liu-chunmei could you upstream the fix of
@tchaikov upstream to seastar or ceph? I don't think seastar will accept this modification. it is only for our special case.

liu-chunmei commented:
> @wjwithagen #33657

@tchaikov is this PR for clang build error?   On my local branch cmake finish success, but seems some files missing changes for CephContext.

wjwithagen commented:
@liu-chunmei 
I needed to revert this PR when compiling under Clang on FreeBSD.
Clang has proven to be picky in some corners of C++ now and then.

tchaikov commented:
> > @wjwithagen #33657
> 
> @tchaikov is this PR for clang build error? On my local branch cmake finish success, but seems some files missing changes for CephContext.

yes.
